---
title: "File conversion"
output: 
  rmarkdown::html_document:
    css: style.css
vignette: >
  %\VignetteIndexEntry{pcadapt}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
  %\usepackage[utf8]{inputenc}
---

## Tips to get your data ready

pcadapt now accepts .bed files as input without any other file conversion. .bed 
files are binary files that contains the genotypes. As PLINK is really 
convenient to use (it is portable, fast and consistant), developing programs 
that rely on this format seems to be a good choice.

## .vcf to .bed

```{bash, eval=FALSE}
./plink -vcf ../comt.chr06.snp.full.final.vcf --geno 0.05 --make-bed --out populus_chr06
```

## Data imputation

Any genetic data come with missing values scattered all over the genotype 
matrix. Some software do not accept data with missing entries (even if you have 
only one missing value), so the best choice left is to impute them. For this 
purpose, I use Beagle for its portability and its ease of use. But, it would be
utopic to believe that no matter how high is the proportion of missing data, 
Beagle will provide very accurate results. Having that in mind, it is desirable
to filter out sites with too many missing values. Here is an example of how to
remove SNPs with more than 5\% of missing values.

```{bash, eval=FALSE}
./plink -vcf ../comt.chr06.snp.full.final.vcf --geno 0.05 --recode vcf --out populus_chr06
```

This will generate a new .vcf file ready to be used with Beagle.

```{bash, eval=FALSE}
java -jar beagle.21Jan17.6cc.jar gt=plink_linux_x86_64/populus_chr06.vcf out=populus_chr06_beagle
```
