<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Principal Component Analysis for outlier detection — pcadapt • pcadapt</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha256-916EbMg70RQy9LHiGkXzG8hSg9EdNy97GazNG/aiY1w=" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/all.min.css" integrity="sha256-nAmazAk6vS34Xqo0BSrTb+abbtFlgsFK7NKSi6o7Y78=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/v4-shims.min.css" integrity="sha256-6qHlizsOWFskGlwVOKuns+D1nB6ssZrHQrNj1wGplHc=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/headroom.min.js" integrity="sha256-DJFC1kqIhelURkuza0AvYal5RxMtpzLjFhsnVIeuk+U=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="Principal Component Analysis for outlier detection — pcadapt" />
<meta property="og:description" content="pcadapt performs principal component analysis and computes p-values to
test for outliers. The test for outliers is based on the correlations between
genetic variation and the first K principal components. pcadapt
also handles Pool-seq data for which the statistical analysis is performed on
the genetic markers frequencies. Returns an object of class pcadapt." />
<meta name="twitter:card" content="summary" />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body>
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">pcadapt</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">4.3.2</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/pcadapt.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/bcm-uga/pcadapt">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Principal Component Analysis for outlier detection</h1>
    <small class="dont-index">Source: <a href='https://github.com/bcm-uga/pcadapt/blob/master/R/pcadapt.R'><code>R/pcadapt.R</code></a></small>
    <div class="hidden name"><code>pcadapt.Rd</code></div>
    </div>

    <div class="ref-description">
    <p><code>pcadapt</code> performs principal component analysis and computes p-values to
test for outliers. The test for outliers is based on the correlations between
genetic variation and the first <code>K</code> principal components. <code>pcadapt</code>
also handles Pool-seq data for which the statistical analysis is performed on
the genetic markers frequencies. Returns an object of class <code>pcadapt</code>.</p>
    </div>

    <pre class="usage"><span class='fu'>pcadapt</span>(
  <span class='no'>input</span>,
  <span class='kw'>K</span> <span class='kw'>=</span> <span class='fl'>2</span>,
  <span class='kw'>method</span> <span class='kw'>=</span> <span class='st'>"mahalanobis"</span>,
  <span class='kw'>min.maf</span> <span class='kw'>=</span> <span class='fl'>0.05</span>,
  <span class='kw'>ploidy</span> <span class='kw'>=</span> <span class='fl'>2</span>,
  <span class='kw'>LD.clumping</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='kw'>pca.only</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>,
  <span class='kw'>tol</span> <span class='kw'>=</span> <span class='fl'>1e-04</span>
)

<span class='co'># S3 method for pcadapt_matrix</span>
<span class='fu'>pcadapt</span>(
  <span class='no'>input</span>,
  <span class='kw'>K</span> <span class='kw'>=</span> <span class='fl'>2</span>,
  <span class='kw'>method</span> <span class='kw'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span>(<span class='st'>"mahalanobis"</span>, <span class='st'>"componentwise"</span>),
  <span class='kw'>min.maf</span> <span class='kw'>=</span> <span class='fl'>0.05</span>,
  <span class='kw'>ploidy</span> <span class='kw'>=</span> <span class='fl'>2</span>,
  <span class='kw'>LD.clumping</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='kw'>pca.only</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>,
  <span class='kw'>tol</span> <span class='kw'>=</span> <span class='fl'>1e-04</span>
)

<span class='co'># S3 method for pcadapt_bed</span>
<span class='fu'>pcadapt</span>(
  <span class='no'>input</span>,
  <span class='kw'>K</span> <span class='kw'>=</span> <span class='fl'>2</span>,
  <span class='kw'>method</span> <span class='kw'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span>(<span class='st'>"mahalanobis"</span>, <span class='st'>"componentwise"</span>),
  <span class='kw'>min.maf</span> <span class='kw'>=</span> <span class='fl'>0.05</span>,
  <span class='kw'>ploidy</span> <span class='kw'>=</span> <span class='fl'>2</span>,
  <span class='kw'>LD.clumping</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='kw'>pca.only</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>,
  <span class='kw'>tol</span> <span class='kw'>=</span> <span class='fl'>1e-04</span>
)

<span class='co'># S3 method for pcadapt_pool</span>
<span class='fu'>pcadapt</span>(
  <span class='no'>input</span>,
  <span class='kw'>K</span> <span class='kw'>=</span> (<span class='fu'><a href='https://rdrr.io/r/base/nrow.html'>nrow</a></span>(<span class='no'>input</span>) - <span class='fl'>1</span>),
  <span class='kw'>method</span> <span class='kw'>=</span> <span class='st'>"mahalanobis"</span>,
  <span class='kw'>min.maf</span> <span class='kw'>=</span> <span class='fl'>0.05</span>,
  <span class='kw'>ploidy</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='kw'>LD.clumping</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='kw'>pca.only</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>,
  <span class='no'>tol</span>
)</pre>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>input</th>
      <td><p>The output of function <code>read.pcadapt</code>.</p></td>
    </tr>
    <tr>
      <th>K</th>
      <td><p>an integer specifying the number of principal components to retain.</p></td>
    </tr>
    <tr>
      <th>method</th>
      <td><p>a character string specifying the method to be used to compute
the p-values. Two statistics are currently available, <code>"mahalanobis"</code>,
and <code>"componentwise"</code>.</p></td>
    </tr>
    <tr>
      <th>min.maf</th>
      <td><p>Threshold of minor allele frequencies above which p-values are 
computed. Default is <code>0.05</code>.</p></td>
    </tr>
    <tr>
      <th>ploidy</th>
      <td><p>Number of trials, parameter of the binomial distribution. 
Default is 2, which corresponds to diploidy, such as for the human genome.</p></td>
    </tr>
    <tr>
      <th>LD.clumping</th>
      <td><p>Default is <code>NULL</code> and doesn't use any SNP thinning.
If you want to use SNP thinning, provide a named list with parameters 
<code>$size</code> and <code>$thr</code> which corresponds respectively to the window 
radius and the squared correlation threshold. A good default value would 
be <code><a href='https://rdrr.io/r/base/list.html'>list(size = 500, thr = 0.1)</a></code>.</p></td>
    </tr>
    <tr>
      <th>pca.only</th>
      <td><p>a logical value indicating whether PCA results should be 
returned (before computing any statistic).</p></td>
    </tr>
    <tr>
      <th>tol</th>
      <td><p>Convergence criterion of <code><a href='https://rdrr.io/pkg/RSpectra/man/svds.html'>RSpectra::svds()</a></code>. 
Default is <code>1e-4</code>.</p></td>
    </tr>
    </table>

    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>The returned value is an object of class <code>pcadapt</code>.</p>
    <h2 class="hasAnchor" id="details"><a class="anchor" href="#details"></a>Details</h2>

    <p>First, a principal component analysis is performed on the scaled and 
centered genotype data. Depending on the specified <code>method</code>, different 
test  statistics can be used.</p>
<p><code>mahalanobis</code> (default): the robust Mahalanobis distance is computed for 
each genetic marker using a robust estimate of both mean and covariance 
matrix between the <code>K</code> vectors of z-scores.</p>
<p><code>communality</code>: the communality statistic measures the proportion of 
variance explained by the first <code>K</code> PCs. Deprecated in version 4.0.0.</p>
<p><code>componentwise</code>: returns a matrix of z-scores.</p>
<p>To compute p-values, test statistics (<code>stat</code>) are divided by a genomic 
inflation factor (<code>gif</code>) when <code>method="mahalanobis"</code>. When using 
<code>method="mahalanobis"</code>, the scaled statistics 
(<code>chi2_stat</code>) should follow a chi-squared distribution with <code>K</code> 
degrees of freedom. When using <code>method="componentwise"</code>, the z-scores 
should follow a chi-squared distribution with <code>1</code> degree of freedom. For 
Pool-seq data, <code>pcadapt</code> provides p-values based on the Mahalanobis 
distance for each SNP.</p>

  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#arguments">Arguments</a></li>
      <li><a href="#value">Value</a></li>
      <li><a href="#details">Details</a></li>
    </ul>

  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by Keurcien Luu, Michael Blum, Florian Privé.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.4.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


