# Generated by using Rcpp::compileAttributes() -> do not edit by hand
# Generator token: 10BE3573-1514-4C36-9D1C-5A225CD40393

colMedian_cpp <- function(x) {
    .Call('pcadapt_colMedian_cpp', PACKAGE = 'pcadapt', x)
}

Erho_cpp <- function(b) {
    .Call('pcadapt_Erho_cpp', PACKAGE = 'pcadapt', b)
}

Es2_cpp <- function(c) {
    .Call('pcadapt_Es2_cpp', PACKAGE = 'pcadapt', c)
}

scaleTau2_matrix <- function(x) {
    .Call('pcadapt_scaleTau2_matrix', PACKAGE = 'pcadapt', x)
}

scaleTau2_vector <- function(x) {
    .Call('pcadapt_scaleTau2_vector', PACKAGE = 'pcadapt', x)
}

covGK_cpp <- function(x, y) {
    .Call('pcadapt_covGK_cpp', PACKAGE = 'pcadapt', x, y)
}

ogk_step <- function(x) {
    .Call('pcadapt_ogk_step', PACKAGE = 'pcadapt', x)
}

getDistance_cpp <- function(x, center, cov) {
    .Call('pcadapt_getDistance_cpp', PACKAGE = 'pcadapt', x, center, cov)
}

#' Robust estimates for location and scatter
#'
#' \code{covRob_cpp} implements the Orthogonalized Gnanadesikan-Kettenring estimator
#' of Maronna and Zamar.
#'
#' @param x a numeric matrix.
#' 
#' @examples
#' ## see also ?pcadapt for examples
#'
#' @export
#'
covRob_cpp <- function(x) {
    .Call('pcadapt_covRob_cpp', PACKAGE = 'pcadapt', x)
}

#' Summary
#'
#' \code{print_convert} prints out a summary of the file conversion.
#'
#' @param input a genotype matrix or a character string specifying the name of the file to be converted.
#' @param output a character string specifying the name of the output file.
#' @param M an integer specifying the number of genetic markers present in the data.
#' @param N an integer specifying the number of individuals present in the data.
#' @param pool an integer specifying the type of data. `0` for genotype data, `1` for pooled data.
#'
#' @examples
#' ## see also ?pcadapt for examples
#'
#' @export
#'
print_convert <- function(input, output, M, N, pool) {
    invisible(.Call('pcadapt_print_convert', PACKAGE = 'pcadapt', input, output, M, N, pool))
}

#' Convert ped files
#'
#' \code{ped2pcadapt} converts \code{ped} files to the format \code{pcadapt}.
#'
#' @param input a character string specifying the name of the file to be converted.
#' @param output a character string specifying the name of the output file.
#' 
#' @examples
#' ## see also ?pcadapt for examples
#'
#' @keywords internal
#'
#' @export
#'
ped2pcadapt <- function(input, output) {
    .Call('pcadapt_ped2pcadapt', PACKAGE = 'pcadapt', input, output)
}

#' Convert lfmm files
#'
#' \code{lfmm2pcadapt} converts \code{lfmm} files to the format \code{pcadapt}.
#'
#' @param input a character string specifying the name of the file to be converted.
#' @param output a character string specifying the name of the output file.
#' 
#' @examples
#' ## see also ?pcadapt for examples
#'
#' @keywords internal
#'
#' @export
#'
lfmm2pcadapt <- function(input, output) {
    .Call('pcadapt_lfmm2pcadapt', PACKAGE = 'pcadapt', input, output)
}

#' Sample genotype matrix from pooled samples
#' 
#' \code{sample_geno_matrix} sample genotypes based on observed allelic frequencies.
#' 
#' @param freq a matrix containing allele frequencies.
#' @param ploidy an integer specifying the ploidy of the sampled individuals.
#' @param sample_size a vector specifying the number of individuals to be sampled for each pool.
#' 
#' @return The returned value is a numeric vector of length 2.
#' 
#' @export
#' 
sample_geno_matrix <- function(freq, ploidy, sample_size) {
    .Call('pcadapt_sample_geno_matrix', PACKAGE = 'pcadapt', freq, ploidy, sample_size)
}

#' File size
#' 
#' \code{get_size_cpp} returns the number of genetic markers and the number of individuals present in the data.
#' 
#' @param filename a character string specifying the name of the file to be processed with \code{pcadapt}.
#' 
#' @return The returned value is a numeric vector of length 2.
#' 
#' @export
#' 
get_size_cpp <- function(filename) {
    .Call('pcadapt_get_size_cpp', PACKAGE = 'pcadapt', filename)
}

#' Covariance for loaded genotype data
#' 
#' \code{cmpt_cov_cpp} computes the covariance matrix of a genotype matrix.
#' 
#' @param filename a character string specifying the name of the file to be processed with \code{pcadapt}.
#' @param xmatrix a genotype matrix.
#' @param min_maf a value between \code{0} and \code{0.45} specifying the threshold of minor allele frequencies above which p-values are computed.
#' @param ploidy an integer specifying the ploidy of the individuals.
#' @param type an integer specifying the input type.
#' 
#' @return The returned value is a Rcpp::List containing the covariance matrix, the number of individuals and the number of genetic markers present in the data.
#' 
#' @export
#' 
cmpt_cov_cpp <- function(filename, xmatrix, min_maf, ploidy, type) {
    .Call('pcadapt_cmpt_cov_cpp', PACKAGE = 'pcadapt', filename, xmatrix, min_maf, ploidy, type)
}

#' Linear regression
#' 
#' \code{lrfunc_cpp} performs the multiple linear regression of the genotype matrix on the scores.
#' 
#' @param filename a character string specifying the name of the file to be processed with \code{pcadapt}.
#' @param xmatrix a genotype matrix.
#' @param scores a matrix containing the scores.
#' @param nIND an integer specifying the number of individuals present in the data.
#' @param nSNP an integer specifying the number of genetic markers present in the data.
#' @param K an integer specifying the number of principal components to retain.
#' @param ploidy an integer specifying the ploidy of the individuals.
#' @param min_maf a value between \code{0} and \code{0.45} specifying the threshold of minor allele frequencies above which p-values are computed.
#' @param sigma a numeric vector.
#' @param type an integer specifying the input type.
#' 
#' @return The returned value is a Rcpp::List containing the multiple linear regression z-scores, the minor allele frequencies and the number of missing values for each genetic marker.
#' 
#' @export
#' 
lrfunc_cpp <- function(filename, xmatrix, scores, nIND, nSNP, K, ploidy, min_maf, sigma, type) {
    .Call('pcadapt_lrfunc_cpp', PACKAGE = 'pcadapt', filename, xmatrix, scores, nIND, nSNP, K, ploidy, min_maf, sigma, type)
}

#' Loadings
#' 
#' \code{cmpt_loadings} computes the loadings.
#' 
#' @param filename a character string specifying the name of the file to be processed with \code{pcadapt}.
#' @param xmatrix a genotype matrix.
#' @param scores a matrix containing the scores.
#' @param nIND an integer specifying the number of individuals present in the data.
#' @param nSNP an integer specifying the number of genetic markers present in the data.
#' @param K an integer specifying the number of principal components to retain.
#' @param ploidy an integer specifying the ploidy of the individuals.
#' @param min_maf a value between \code{0} and \code{0.45} specifying the threshold of minor allele frequencies above which p-values are computed.
#' @param sigma a numeric vector.
#' @param type an integer specifying the input type.
#' 
#' @return The returned value is a Rcpp::List containing the multiple linear regression z-scores, the minor allele frequencies and the number of missing values for each genetic marker.
#' 
#' @export
#' 
cmpt_loadings <- function(filename, xmatrix, scores, nIND, nSNP, K, ploidy, min_maf, sigma, type) {
    .Call('pcadapt_cmpt_loadings', PACKAGE = 'pcadapt', filename, xmatrix, scores, nIND, nSNP, K, ploidy, min_maf, sigma, type)
}

#' Sample genotype matrix from pooled samples
#' 
#' \code{sample_geno_file} sample genotypes based on observed allelic frequencies.
#' 
#' @param input a character string specifying the name of the file containing the allele frequencies.
#' @param output a character string specifying the name of the output file.
#' @param ploidy an integer specifying the ploidy of the sampled individuals.
#' @param sample_size a vector specifying the number of individuals to be sampled for each pool.
#' 
#' @return The returned value is a numeric vector of length 2.
#' 
#' @export
#' 
sample_geno_file <- function(input, output, ploidy, sample_size) {
    .Call('pcadapt_sample_geno_file', PACKAGE = 'pcadapt', input, output, ploidy, sample_size)
}

get_geno_char <- function(allele_sep) {
    .Call('pcadapt_get_geno_char', PACKAGE = 'pcadapt', allele_sep)
}

get_geno_int <- function(allele_sep) {
    .Call('pcadapt_get_geno_int', PACKAGE = 'pcadapt', allele_sep)
}

check_line_na <- function(string_geno_row, geno_char) {
    .Call('pcadapt_check_line_na', PACKAGE = 'pcadapt', string_geno_row, geno_char)
}

#' Convert vcfR genotype matrices
#'
#' \code{vcf_convert} converts outputs of \code{extract.gt} to the format \code{pcadapt}.
#'
#' @param string_geno a genotype matrix extracted from a VCF file with `vcfR`. 
#' @param output a character string indicating the name of the output file.
#' @param allele.sep a vector of characters indicating what delimiters are used to separate alleles.
#'
#' @examples
#' ## see also ?pcadapt for examples
#'
#' @keywords internal
#'
#' @export
#'
vcf_convert <- function(string_geno, output, allele_sep) {
    .Call('pcadapt_vcf_convert', PACKAGE = 'pcadapt', string_geno, output, allele_sep)
}

