# Generated by using Rcpp::compileAttributes() -> do not edit by hand
# Generator token: 10BE3573-1514-4C36-9D1C-5A225CD40393

#' File size
#' 
#' \code{get_size_file} returns the number of genetic markers and the number of individuals present in the data.
#' 
#' @param path a character string specifying the name of the file to be processed with \code{pcadapt}.
#' 
#' @return The returned value is a numeric vector of length 2.
#' 
#' @export
#' 
get_size_file <- function(path) {
    .Call('pcadapt_get_size_file', PACKAGE = 'pcadapt', path)
}

#' Covariance for genotype data stored in an external file
#' 
#' \code{cmpt_cov_file} computes the covariance matrix of a genotype matrix when the genotype matrix is stored in an external file.
#' 
#' @param path a character string specifying the name of the file to be processed with \code{pcadapt}.
#' @param min_maf a value between \code{0} and \code{0.45} specifying the threshold of minor allele frequencies above which p-values are computed.
#' @param ploidy an integer specifying the ploidy of the individuals.
#' 
#' @return The returned value is a Rcpp::List containing the covariance matrix, the number of individuals and the number of genetic markers present in the data.
#' 
#' @export
#' 
cmpt_cov_file <- function(path, min_maf, ploidy) {
    .Call('pcadapt_cmpt_cov_file', PACKAGE = 'pcadapt', path, min_maf, ploidy)
}

#' Linear regression
#' 
#' \code{lrfunc_file} performs the multiple linear regression of the genotype matrix on the scores when the genotype matrix is stored in an external file.
#' 
#' @param filename a character string specifying the name of the file to be processed with \code{pcadapt}.
#' @param scores a matrix containing the scores.
#' @param nIND an integer specifying the number of individuals present in the data.
#' @param nSNP an integer specifying the number of genetic markers present in the data.
#' @param K an integer specifying the number of principal components to retain.
#' @param ploidy an integer specifying the ploidy of the individuals.
#' @param min_maf a value between \code{0} and \code{0.45} specifying the threshold of minor allele frequencies above which p-values are computed.
#' 
#' @return The returned value is a Rcpp::List containing the multiple linear regression z-scores, the minor allele frequencies and the number of missing values for each genetic marker.
#' 
#' @export
#' 
lrfunc_file <- function(filename, scores, nIND, nSNP, K, ploidy, min_maf) {
    .Call('pcadapt_lrfunc_file', PACKAGE = 'pcadapt', filename, scores, nIND, nSNP, K, ploidy, min_maf)
}

#' Sample genotype matrix from pooled samples
#' 
#' \code{sample_geno_file} sample genotypes based on observed allelic frequencies.
#' 
#' @param input a character string specifying the name of the file containing the allele frequencies.
#' @param output a character string specifying the name of the output file.
#' @param ploidy an integer specifying the ploidy of the sampled individuals.
#' @param sample_size a vector specifying the number of individuals to be sampled for each pool.
#' 
#' @return The returned value is a numeric vector of length 2.
#' 
#' @export
#' 
sample_geno_file <- function(input, output, ploidy, sample_size) {
    .Call('pcadapt_sample_geno_file', PACKAGE = 'pcadapt', input, output, ploidy, sample_size)
}

#' Product of a matrix with its transpose
#' 
#' \code{tAA_cpp} computes the product of a real-valued matrix with its transpose. 
#' 
#' @param x a matrix.
#' @param nrow an integer specifying the number of rows of \code{x}.
#' @param ncol an integer specifying the number of columns of \code{x}.
#' 
#' @return The returned value is the result of the product of \code{x} with its transpose.
#' 
#' @export
#' 
tAA_cpp <- function(x, nrow, ncol) {
    .Call('pcadapt_tAA_cpp', PACKAGE = 'pcadapt', x, nrow, ncol)
}

#' Covariance for loaded genotype data
#' 
#' \code{cmpt_cov_file} computes the covariance matrix of a genotype matrix.
#' 
#' @param input a genotype matrix.
#' @param min_maf a value between \code{0} and \code{0.45} specifying the threshold of minor allele frequencies above which p-values are computed.
#' @param ploidy an integer specifying the ploidy of the individuals.
#' 
#' @return The returned value is a Rcpp::List containing the covariance matrix, the number of individuals and the number of genetic markers present in the data.
#' 
#' @export
#' 
cmpt_cov_matrix <- function(input, min_maf, ploidy) {
    .Call('pcadapt_cmpt_cov_matrix', PACKAGE = 'pcadapt', input, min_maf, ploidy)
}

#' Linear regression
#' 
#' \code{lrfunc_matrix} performs the multiple linear regression of the genotype matrix on the scores.
#' 
#' @param Geno a genotype matrix.
#' @param scores a matrix containing the scores.
#' @param nIND an integer specifying the number of individuals present in the data.
#' @param nSNP an integer specifying the number of genetic markers present in the data.
#' @param K an integer specifying the number of principal components to retain.
#' @param ploidy an integer specifying the ploidy of the individuals.
#' @param min_maf a value between \code{0} and \code{0.45} specifying the threshold of minor allele frequencies above which p-values are computed.
#' 
#' @return The returned value is a Rcpp::List containing the multiple linear regression z-scores, the minor allele frequencies and the number of missing values for each genetic marker.
#' 
#' @export
#' 
lrfunc_matrix <- function(Geno, scores, nIND, nSNP, K, ploidy, min_maf) {
    .Call('pcadapt_lrfunc_matrix', PACKAGE = 'pcadapt', Geno, scores, nIND, nSNP, K, ploidy, min_maf)
}

#' Convert ped files
#'
#' \code{ped2pcadapt} converts \code{ped} files to the format \code{pcadapt}.
#'
#' @param input a character string specifying the name of the file to be converted.
#' @param output a character string specifying the name of the output file.
#' 
#' @examples
#' ## see also ?pcadapt for examples
#'
#' @keywords internal
#'
#' @export
#'
ped2pcadapt <- function(input, output) {
    .Call('pcadapt_ped2pcadapt', PACKAGE = 'pcadapt', input, output)
}

#' Convert lfmm files
#'
#' \code{lfmm2pcadapt} converts \code{lfmm} files to the format \code{pcadapt}.
#'
#' @param input a character string specifying the name of the file to be converted.
#' @param output a character string specifying the name of the output file.
#' 
#' @examples
#' ## see also ?pcadapt for examples
#'
#' @keywords internal
#'
#' @export
#'
lfmm2pcadapt <- function(input, output) {
    .Call('pcadapt_lfmm2pcadapt', PACKAGE = 'pcadapt', input, output)
}

#' Sample genotype matrix from pooled samples
#' 
#' \code{sample_geno_matrix} sample genotypes based on observed allelic frequencies.
#' 
#' @param freq a matrix containing allele frequencies.
#' @param ploidy an integer specifying the ploidy of the sampled individuals.
#' @param sample_size a vector specifying the number of individuals to be sampled for each pool.
#' 
#' @return The returned value is a numeric vector of length 2.
#' 
#' @export
#' 
sample_geno_matrix <- function(freq, ploidy, sample_size) {
    .Call('pcadapt_sample_geno_matrix', PACKAGE = 'pcadapt', freq, ploidy, sample_size)
}

get_geno_char <- function(allele_sep) {
    .Call('pcadapt_get_geno_char', PACKAGE = 'pcadapt', allele_sep)
}

get_geno_int <- function(allele_sep) {
    .Call('pcadapt_get_geno_int', PACKAGE = 'pcadapt', allele_sep)
}

check_line_na <- function(string_geno_row, geno_char) {
    .Call('pcadapt_check_line_na', PACKAGE = 'pcadapt', string_geno_row, geno_char)
}

#' Convert vcfR genotype matrices
#'
#' \code{vcf_convert} converts outputs of \code{extract.gt} to the format \code{pcadapt}.
#'
#' @param string_geno a genotype matrix extracted from a VCF file with `vcfR`. 
#' @param output a character string indicating the name of the output file.
#' @param allele.sep a vector of characters indicating what delimiters are used to separate alleles.
#'
#' @examples
#' ## see also ?pcadapt for examples
#'
#' @keywords internal
#'
#' @export
#'
vcf_convert <- function(string_geno, output, allele_sep) {
    .Call('pcadapt_vcf_convert', PACKAGE = 'pcadapt', string_geno, output, allele_sep)
}

