# Generated by using Rcpp::compileAttributes() -> do not edit by hand
# Generator token: 10BE3573-1514-4C36-9D1C-5A225CD40393

#' Median
#' 
#' \code{median_row_i} computes the median of a specific row.
#' 
#' @param x a genotype matrix.
#' @param i an integer.
#' 
#' @return The returned value is a real number.
#' 
#' @export
#' 
median_row_i <- function(x, i) {
    .Call('pcadapt_median_row_i', PACKAGE = 'pcadapt', x, i)
}

#' Median
#' 
#' \code{median_per_pop} computes the median of each population for a specific row.
#' 
#' @param x a genotype matrix.
#' @param lab a vector of integers.
#' @param i an integer.
#' 
#' @return The returned value is a real-valued vector .
#' 
#' @export
#' 
median_per_pop <- function(x, lab, pop, i) {
    .Call('pcadapt_median_per_pop', PACKAGE = 'pcadapt', x, lab, pop, i)
}

#' @export
#' 
check_row <- function(x, i) {
    .Call('pcadapt_check_row', PACKAGE = 'pcadapt', x, i)
}

#' @export
#' 
impute_geno <- function(x) {
    .Call('pcadapt_impute_geno', PACKAGE = 'pcadapt', x)
}

#' @export
#' 
impute_geno_pop <- function(x, lab, pop) {
    .Call('pcadapt_impute_geno_pop', PACKAGE = 'pcadapt', x, lab, pop)
}

#' Number of individuals in a specific population
#' 
#' \code{get_nb_ind} returns the number of individuals in a specific population.
#' 
#' @param lab a vector of integers.
#' @param anc an integer.
#' 
#' @return The returned value is an integer.
#' 
#' @export
#' 
get_nb_ind <- function(lab, anc) {
    .Call('pcadapt_get_nb_ind', PACKAGE = 'pcadapt', lab, anc)
}

#' @export
#' 
cmpt_centroids <- function(u, lab, anc1, anc2) {
    .Call('pcadapt_cmpt_centroids', PACKAGE = 'pcadapt', u, lab, anc1, anc2)
}

#' @export
#' 
cmpt_transformation <- function(uloc, uglob, lab, ancstrl1, ancstrl2, s, dloc, dglob) {
    invisible(.Call('pcadapt_cmpt_transformation', PACKAGE = 'pcadapt', uloc, uglob, lab, ancstrl1, ancstrl2, s, dloc, dglob))
}

#' @export
#' 
rescale_local_pca <- function(u, s, dep_loc, dep_glob) {
    .Call('pcadapt_rescale_local_pca', PACKAGE = 'pcadapt', u, s, dep_loc, dep_glob)
}

#' Global Principal Component Analysis
#' 
#' \code{cmpt_global_pca} computes the scores using all genetic markers.
#' 
#' @param geno a genotype matrix.
#' @param V a loading matrix.
#' @param sigma a vector of singular values.
#' 
#' @return The returned value is a matrix of scores.
#' 
#' @export
#' 
cmpt_global_pca <- function(geno, V, sigma) {
    .Call('pcadapt_cmpt_global_pca', PACKAGE = 'pcadapt', geno, V, sigma)
}

#' Local Principal Component Analysis
#' 
#' \code{cmpt_local_pca} computes the scores using a subset of genetic markers.
#' 
#' @param geno a genotype matrix.
#' @param V a loading matrix.
#' @param sigma a vector of singular values.
#' @param beg an integer specifying the first marker to be included.
#' @param end an integer specifying the first marker to be excluded.
#' 
#' @return The returned value is a matrix of scores.
#' 
#' @export
#' 
cmpt_local_pca <- function(geno, V, sigma, beg, end) {
    .Call('pcadapt_cmpt_local_pca', PACKAGE = 'pcadapt', geno, V, sigma, beg, end)
}

#' Update local Principal Component Analysis
#' 
#' \code{updt_local_scores} computes the scores using a subset of genetic markers.
#' 
#' @param u a score matrix.
#' @param geno a genotype matrix.
#' @param V a loading matrix.
#' @param sigma a vector of singular values.
#' @param beg an integer specifying the first marker to be included.
#' @param end an integer specifying the first marker to be excluded.
#' 
#' @return The returned value is a score matrix.
#' 
#' @export
#' 
updt_local_scores <- function(u, geno, V, sigma, beg, end) {
    invisible(.Call('pcadapt_updt_local_scores', PACKAGE = 'pcadapt', u, geno, V, sigma, beg, end))
}

#' @export
#' 
cmpt_window_stat <- function(uloc, uglob, direction, lab, adm, axis) {
    .Call('pcadapt_cmpt_window_stat', PACKAGE = 'pcadapt', uloc, uglob, direction, lab, adm, axis)
}

#' @export
#' 
get_rank <- function(v_temp) {
    .Call('pcadapt_get_rank', PACKAGE = 'pcadapt', v_temp)
}

#' @export
#' 
cmpt_window_wilcoxon <- function(uloc, uglob, direction, lab, adm, axis) {
    .Call('pcadapt_cmpt_window_wilcoxon', PACKAGE = 'pcadapt', uloc, uglob, direction, lab, adm, axis)
}

#' @export
#' 
cmpt_all_stat <- function(geno, V, sigma, window_size, direction, lab, ancstrl1, ancstrl2, adm, axis) {
    .Call('pcadapt_cmpt_all_stat', PACKAGE = 'pcadapt', geno, V, sigma, window_size, direction, lab, ancstrl1, ancstrl2, adm, axis)
}

colMedian_cpp <- function(x) {
    .Call('pcadapt_colMedian_cpp', PACKAGE = 'pcadapt', x)
}

Erho_cpp <- function(b) {
    .Call('pcadapt_Erho_cpp', PACKAGE = 'pcadapt', b)
}

Es2_cpp <- function(c) {
    .Call('pcadapt_Es2_cpp', PACKAGE = 'pcadapt', c)
}

scaleTau2_matrix <- function(x) {
    .Call('pcadapt_scaleTau2_matrix', PACKAGE = 'pcadapt', x)
}

scaleTau2_vector <- function(x) {
    .Call('pcadapt_scaleTau2_vector', PACKAGE = 'pcadapt', x)
}

covGK_cpp <- function(x, y) {
    .Call('pcadapt_covGK_cpp', PACKAGE = 'pcadapt', x, y)
}

ogk_step <- function(x) {
    .Call('pcadapt_ogk_step', PACKAGE = 'pcadapt', x)
}

getDistance_cpp <- function(x, center, cov) {
    .Call('pcadapt_getDistance_cpp', PACKAGE = 'pcadapt', x, center, cov)
}

#' Robust estimates for location and scatter
#'
#' \code{covRob_cpp} implements the Orthogonalized Gnanadesikan-Kettenring estimator
#' of Maronna and Zamar.
#'
#' @param x a numeric matrix.
#' 
#' @examples
#' ## see also ?pcadapt for examples
#'
#' @export
#'
covRob_cpp <- function(x) {
    .Call('pcadapt_covRob_cpp', PACKAGE = 'pcadapt', x)
}

#' Summary
#'
#' \code{print_convert} prints out a summary of the file conversion.
#'
#' @param input a genotype matrix or a character string specifying the name of the file to be converted.
#' @param output a character string specifying the name of the output file.
#' @param M an integer specifying the number of genetic markers present in the data.
#' @param N an integer specifying the number of individuals present in the data.
#' @param pool an integer specifying the type of data. `0` for genotype data, `1` for pooled data.
#'
#' @examples
#' ## see also ?pcadapt for examples
#'
#' @export
#'
print_convert <- function(input, output, M, N, pool) {
    invisible(.Call('pcadapt_print_convert', PACKAGE = 'pcadapt', input, output, M, N, pool))
}

#' Convert ped files
#'
#' \code{ped2pcadapt} converts \code{ped} files to the format \code{pcadapt}.
#'
#' @param input a character string specifying the name of the file to be converted.
#' @param output a character string specifying the name of the output file.
#' 
#' @examples
#' ## see also ?pcadapt for examples
#'
#' @keywords internal
#'
#' @export
#'
ped2pcadapt <- function(input, output) {
    .Call('pcadapt_ped2pcadapt', PACKAGE = 'pcadapt', input, output)
}

#' Convert lfmm files
#'
#' \code{lfmm2pcadapt} converts \code{lfmm} files to the format \code{pcadapt}.
#'
#' @param input a character string specifying the name of the file to be converted.
#' @param output a character string specifying the name of the output file.
#' 
#' @examples
#' ## see also ?pcadapt for examples
#'
#' @keywords internal
#'
#' @export
#'
lfmm2pcadapt <- function(input, output) {
    .Call('pcadapt_lfmm2pcadapt', PACKAGE = 'pcadapt', input, output)
}

#' Sample genotype matrix from pooled samples
#' 
#' \code{sample_geno_matrix} sample genotypes based on observed allelic frequencies.
#' 
#' @param freq a matrix containing allele frequencies.
#' @param ploidy an integer specifying the ploidy of the sampled individuals.
#' @param sample_size a vector specifying the number of individuals to be sampled for each pool.
#' 
#' @return The returned value is a numeric vector of length 2.
#' 
#' @export
#' 
sample_geno_matrix <- function(freq, ploidy, sample_size) {
    .Call('pcadapt_sample_geno_matrix', PACKAGE = 'pcadapt', freq, ploidy, sample_size)
}

#' File size
#' 
#' \code{get_size_cpp} returns the number of genetic markers and the number of individuals present in the data.
#' 
#' @param filename a character string specifying the name of the file to be processed with \code{pcadapt}.
#' 
#' @return The returned value is a numeric vector of length 2.
#' 
#' @export
#' 
get_size_cpp <- function(filename) {
    .Call('pcadapt_get_size_cpp', PACKAGE = 'pcadapt', filename)
}

#' @export
#' 
cmpt_minor_af <- function(xmatrix, ploidy) {
    .Call('pcadapt_cmpt_minor_af', PACKAGE = 'pcadapt', xmatrix, ploidy)
}

#' @export
#' 
scale_geno <- function(xmatrix, ploidy, maf, keep_or_not) {
    .Call('pcadapt_scale_geno', PACKAGE = 'pcadapt', xmatrix, ploidy, maf, keep_or_not)
}

#' Covariance for loaded genotype data
#' 
#' \code{cmpt_cov_cpp} computes the covariance matrix of a genotype matrix.
#' 
#' @param filename a character string specifying the name of the file to be processed with \code{pcadapt}.
#' @param xmatrix a genotype matrix.
#' @param min_maf a value between \code{0} and \code{0.45} specifying the threshold of minor allele frequencies above which p-values are computed.
#' @param ploidy an integer specifying the ploidy of the individuals.
#' @param type an integer specifying the input type.
#' 
#' @return The returned value is a Rcpp::List containing the covariance matrix, the number of individuals and the number of genetic markers present in the data.
#' 
#' @export
#' 
cmpt_cov_cpp <- function(filename, xmatrix, min_maf, ploidy, type) {
    .Call('pcadapt_cmpt_cov_cpp', PACKAGE = 'pcadapt', filename, xmatrix, min_maf, ploidy, type)
}

#' Compute the loadings
#' 
#' \code{cmpt_loadings} returns the loadings.
#' 
#' @param filename a character string specifying the name of the file to be processed with \code{pcadapt}.
#' @param xmatrix a genotype matrix.
#' @param scores a matrix containing the scores.
#' @param nIND an integer specifying the number of individuals present in the data.
#' @param nSNP an integer specifying the number of genetic markers present in the data.
#' @param K an integer specifying the number of principal components to retain.
#' @param ploidy an integer specifying the ploidy of the individuals.
#' @param min_maf a value between \code{0} and \code{0.45} specifying the threshold of minor allele frequencies above which p-values are computed.
#' @param sigma a numeric vector.
#' @param type an integer specifying the input type.
#' 
#' @return The returned value is a matrix containing the loadings.
#' 
#' @export
#' 
cmpt_loadings <- function(filename, xmatrix, scores, nIND, nSNP, K, ploidy, min_maf, sigma, type) {
    .Call('pcadapt_cmpt_loadings', PACKAGE = 'pcadapt', filename, xmatrix, scores, nIND, nSNP, K, ploidy, min_maf, sigma, type)
}

#' Linear regression
#' 
#' \code{lrfunc_cpp} performs the multiple linear regression of the genotype matrix on the scores.
#' 
#' @param filename a character string specifying the name of the file to be processed with \code{pcadapt}.
#' @param xmatrix a genotype matrix.
#' @param scores a matrix containing the scores.
#' @param nIND an integer specifying the number of individuals present in the data.
#' @param nSNP an integer specifying the number of genetic markers present in the data.
#' @param K an integer specifying the number of principal components to retain.
#' @param ploidy an integer specifying the ploidy of the individuals.
#' @param min_maf a value between \code{0} and \code{0.45} specifying the threshold of minor allele frequencies above which p-values are computed.
#' @param sigma a numeric vector.
#' @param type an integer specifying the input type.
#' 
#' @return The returned value is a Rcpp::List containing the multiple linear regression z-scores, the minor allele frequencies and the number of missing values for each genetic marker.
#' 
#' @export
#' 
lrfunc_cpp <- function(filename, xmatrix, scores, nIND, nSNP, K, ploidy, min_maf, sigma, type) {
    .Call('pcadapt_lrfunc_cpp', PACKAGE = 'pcadapt', filename, xmatrix, scores, nIND, nSNP, K, ploidy, min_maf, sigma, type)
}

#' Sample genotype matrix from pooled samples
#' 
#' \code{sample_geno_file} sample genotypes based on observed allelic frequencies.
#' 
#' @param input a character string specifying the name of the file containing the allele frequencies.
#' @param output a character string specifying the name of the output file.
#' @param ploidy an integer specifying the ploidy of the sampled individuals.
#' @param sample_size a vector specifying the number of individuals to be sampled for each pool.
#' 
#' @return The returned value is a numeric vector of length 2.
#' 
#' @export
#' 
sample_geno_file <- function(input, output, ploidy, sample_size) {
    .Call('pcadapt_sample_geno_file', PACKAGE = 'pcadapt', input, output, ploidy, sample_size)
}

get_geno_char <- function(allele_sep) {
    .Call('pcadapt_get_geno_char', PACKAGE = 'pcadapt', allele_sep)
}

get_geno_int <- function(allele_sep) {
    .Call('pcadapt_get_geno_int', PACKAGE = 'pcadapt', allele_sep)
}

check_line_na <- function(string_geno_row, geno_char) {
    .Call('pcadapt_check_line_na', PACKAGE = 'pcadapt', string_geno_row, geno_char)
}

#' Convert vcfR genotype matrices
#'
#' \code{vcf_convert} converts outputs of \code{extract.gt} to the format \code{pcadapt}.
#'
#' @param string_geno a genotype matrix extracted from a VCF file with `vcfR`. 
#' @param output a character string indicating the name of the output file.
#' @param allele.sep a vector of characters indicating what delimiters are used to separate alleles.
#'
#' @examples
#' ## see also ?pcadapt for examples
#'
#' @keywords internal
#'
#' @export
#'
vcf_convert <- function(string_geno, output, allele_sep) {
    .Call('pcadapt_vcf_convert', PACKAGE = 'pcadapt', string_geno, output, allele_sep)
}

